src/coreclr/nativeaot/Runtime/GCToEEInterface.cpp

```diff
@@ -625,6 +90,64 @@ void GCToEEInterface::BeforeGcScanRoots(int condemned, bool is_bgc, bool is_conc
 #endif
 }
 
+void GCToEEInterface::GcScanRoots(promote_func* fn, int condemned, int max_gen, ScanContext* sc)
+{
+    // STRESS_LOG1(LF_GCROOTS, LL_INFO10, "GCScan: Phase = %s\n", sc->promotion ? "promote" : "relocate");
+
+    FOREACH_THREAD(pThread)
+    {
+        // Skip "GC Special" threads which are really background workers that will never have any roots.
+        if (pThread->IsGCSpecial())
+            continue;
+
+#if !defined (ISOLATED_HEAPS)
+        if (!GCHeapUtilities::GetGCHeap()->IsThreadUsingAllocationContextHeap(pThread->GetAllocContext(),
+            sc->thread_number))
+        {
+            // STRESS_LOG2(LF_GC|LF_GCROOTS, LL_INFO100, "{ Scan of Thread %p (ID = %x) declined by this heap\n",
+            //             pThread, pThread->GetThreadId());
+        }
+        else
+#endif
+        {
+            InlinedThreadStaticRoot* pRoot = pThread->GetInlinedThreadStaticList();
+            while (pRoot != NULL)
+            {
+                STRESS_LOG2(LF_GC | LF_GCROOTS, LL_INFO100, "{ Scanning Thread's %p inline thread statics root %p. \n", pThread, pRoot);
+                RedhawkGCInterface::EnumGcRef((PTR_RtuObjectRef)&pRoot->m_threadStaticsBase, GCRK_Object, (void*)fn, sc);
```

Comment:
Can we use the actual types in signatures to avoid these casts?

Reply:
It is possible to avoid casts of scan funcs. I was just afraid it would cause too much of cascading changes, but it ended up not too bad.

I think I will change RtuObjectRef --> ObjectRef  as well, while I am at this.   (and just make it an alias of `Object*`)